File: exercises/extra/06_RDataFrame/solutions/02_collections.cpp
/*
C++ solution for the 02_collections exercise, using fully-compiled RDataFrame execution.
*/
#include <string>

#include <TCanvas.h>
#include <ROOT/RDataFrame.hxx>
#include <ROOT/RLogger.hxx>
#include <ROOT/RVec.hxx>

int main()
{
    // Create an RDataFrame to process the dataset
    const std::string dataset_name{"dataset"};
    const std::string file_path{"../data/example_file.root"};
    ROOT::RDataFrame df{dataset_name, file_path};

    // Select the elements of `vec1` for which the square of the corresponding elements in `vec2` is less than 0.1
    // Create the histogram of the new column
    auto h = df.Define("v1_to_plot", [](const ROOT::RVecF &v1, const ROOT::RVecF &v2)
                       { return v1[v2 * v2 < 0.1]; },
                       {"vec1", "vec2"})
                 .Histo1D<ROOT::RVecF>("v1_to_plot");

    // Produce the plot
    TCanvas c;
    h->Draw();
    c.SaveAs("02_collections.png");
}

File: exercises/extra/06_RDataFrame/solutions/02_collections.py
"""
Python solution for the 02_collections exercise using RDataFrame with JIT compiled C++ functions.
"""
import ROOT

# Create an RDataFrame to process the dataset
dataset_name = "dataset"
file_path = "../data/example_file.root"
df = ROOT.RDataFrame(dataset_name, file_path)

# Select the elements of `vec1` for which the square of the corresponding elements in `vec2` is less than 0.1
# Create the histogram of the new column
h = df.Define("v1_to_plot", "vec1[vec2*vec2 < 0.1]").Histo1D("v1_to_plot")

# Produce the plot
c = ROOT.TCanvas()
h.Draw()
c.SaveAs("02_collections.png")
